File: CS261/mylibrary_cpp/library.cpp
#include <cstdarg>		// support macros for vararg
#include <iostream>
#include "library.h"

/* TODO:
 * xAdd book
 * xAdd keywords
 * xAdd music album
 * xAdd band members
 * xAdd movie
 * xAdd cast
 * xPrint items by title
 * xPrint items by keyword
 * xPrint books by author
 * xPrint albums by band
 * xPrint albums by member
 * xPrint movies by director
 * xPrint movies by cast 
 * xDelete set contents 
 * xDelete map contents
 * xFix dependence on casting in keywords
 * Fix dependence on casting in addBandMembers and addCast
 *
 *  Bugs:
 * xItems sorted by title
 * xExtra comma on keywords output
 *  Missing/extra newline when printing items
*/

// general functions
/*Library::Library() // Needed ?
{
    // your code here
}
*/

Library::~Library()
{
    // your code here

    deleteItemSetContents(AlbumSet);
    deleteItemSetContents(BookSet);
    deleteItemSetContents(MovieSet);
    deleteMapContents(ItemMapByKeyword);
//    deleteMapContents(ItemMapByTitle);
    deleteMapContents(AlbumMapByBand);
    deleteMapContents(AlbumMapByMember);
    deleteMapContents(BookMapByAuthor);
    deleteMapContents(MovieMapByCast);
    deleteMapContents(MovieMapByDirector);
}

void Library::addKeywordsForItem(const Item* const item, int nKeywords, ...)
{
    // the code in this function demonstrates how to handle a vararg in C++

    va_list		keywords;
    char		*keyword;

    va_start(keywords, nKeywords);

    // Check AllItemSet for item
    
    Item * nItem;
//    Item * nItem = const_cast<Item*>(item); // Temp fix;

    map<string, ItemPtr>::iterator it;
    it = ItemMapByTitle.find(item->getTitle());
    nItem = it->second.getPtr();

    for (int i = 0; i < nKeywords; i++)
    {
	    keyword = va_arg(keywords, char*);

	    // do something with each keyword
        
        nItem->setKeywords(keyword);

        if (ItemMapByKeyword.find(keyword) == ItemMapByKeyword.end()) // Case 1 - keyword not mapped
        {
            ItemSet * tempSet = new ItemSet();

            tempSet->insert(nItem);
            ItemMapByKeyword.insert(std::make_pair(keyword, tempSet));
        }
        else // Case 2 - keyword already mapped
        {
            ItemSet * tempSet = ItemMapByKeyword[keyword];

            tempSet->insert(nItem);
            ItemMapByKeyword.insert(std::make_pair(keyword, tempSet));
        }
    }

    va_end(keywords);
}

const ItemSet* Library::itemsForKeyword(const string& keyword) const
{
    // your code here
    
    const ItemSet * tempSet;

    if (ItemMapByKeyword.find(keyword) != ItemMapByKeyword.end())
    {
        tempSet = ItemMapByKeyword.at(keyword);

        return tempSet;
    }
    else
    {
        return NULL;
    }
}

// book-related functions

const Item* Library::addBook(const string& title, const string& author, const int nPages)
{
    // your code here

    Book *nBook = new Book(title, author, nPages);

    BookSet.insert(nBook);
    
    if (BookMapByAuthor.find(author) == BookMapByAuthor.end()) // Case 1 - author not mapped
        {

            ItemSet * tempSet = new ItemSet();
            tempSet->insert(nBook);
            BookMapByAuthor.insert(std::make_pair(author, tempSet));
            ItemMapByTitle.insert(std::make_pair(title, nBook));
        }
        else // Case 2 -  already mapped
        {

            ItemSet * tempSet;
            tempSet = BookMapByAuthor[author];
            tempSet->insert(nBook);
            BookMapByAuthor.insert(std::make_pair(author, tempSet));
            ItemMapByTitle.insert(std::make_pair(title, nBook));
        }

    return nBook;
}

const ItemSet* Library::booksByAuthor(const string& author) const
{
    // your code here

    const ItemSet * tempSet;

    if (BookMapByAuthor.find(author) != BookMapByAuthor.end())
    {
        tempSet = BookMapByAuthor.at(author);

        return tempSet;
    }
    else
    {
        return NULL;
    }
}

const ItemSet* Library::books() const
{
    // your code here

    return &BookSet;
}

// music-related functions

const Item* Library::addMusicAlbum(const string& title, const string& band, const int nSongs)
{
    // your code here

    Item *nAlbum = new MusicAlbum(title, band, nSongs);
    AlbumSet.insert(nAlbum);
    
    if (AlbumMapByBand.find(band) == AlbumMapByBand.end()) // Case 1 - band not mapped
        {
            ItemSet * tempSet = new ItemSet();
            tempSet->insert(nAlbum);
            AlbumMapByBand.insert(std::make_pair(band, tempSet));
            ItemMapByTitle.insert(std::make_pair(title, nAlbum));
        }
        else // Case 2 -  already mapped
        {
            ItemSet * tempSet;
            tempSet = AlbumMapByBand[band];
            tempSet->insert(nAlbum);
            AlbumMapByBand.insert(std::make_pair(band, tempSet));
            ItemMapByTitle.insert(std::make_pair(title, nAlbum));
        }

    return nAlbum;
}

void Library::addBandMembers(const Item* const musicAlbum, const int nBandMembers, ...)
{
    // your code here

    va_list members;
    char *member;

    va_start(members, nBandMembers);

    for (int i = 0; i < nBandMembers; i++)
    {
	    member = va_arg(members, char*);

	    // do something with each keyword
        
        MusicAlbum * nMusicAlbum;

        map<string, ItemPtr>::iterator it;
        it = ItemMapByTitle.find(musicAlbum->getTitle());
        nMusicAlbum = (MusicAlbum*)it->second.getPtr(); // Temp fix

        nMusicAlbum->setMembers(member);

        if (AlbumMapByMember.find(member) == AlbumMapByMember.end()) // Case 1 - keyword not mapped
        {
            ItemSet * tempSet = new ItemSet();
            
            tempSet->insert(nMusicAlbum);
            AlbumMapByMember.insert(std::make_pair(member, tempSet));
        }
        else // Case 2 - keyword already mapped
        {
            ItemSet * tempSet = AlbumMapByMember[member];

            tempSet->insert(nMusicAlbum);
            AlbumMapByMember.insert(std::make_pair(member, tempSet));
        }
    }

    va_end(members);
}

const ItemSet* Library::musicByBand(const string& band) const
{
    // your code here
    
    const ItemSet * tempSet;

    if (AlbumMapByBand.find(band) != AlbumMapByBand.end())
    {
        tempSet = AlbumMapByBand.at(band);

        return tempSet;
    }
    else
    {
        return NULL;
    }
}

const ItemSet* Library::musicByMusician(const string& musician) const
{
    // your code here

    const ItemSet * tempSet;

    if (AlbumMapByMember.find(musician) != AlbumMapByMember.end())
    {
        tempSet = AlbumMapByMember.at(musician);

        return tempSet;
    }
    else
    {
        return NULL;
    }
}

const ItemSet* Library::musicAlbums() const
{
    // your code here

    return &AlbumSet;
}

// movie-related functions

const Item* Library::addMovie(const string& title, const string& director, const int nScenes)
{
    // your code here

    Item *nMovie = new Movie(title, director, nScenes);
    MovieSet.insert(nMovie);
    
    if (MovieMapByDirector.find(director) == MovieMapByDirector.end()) // Case 1 - director not mapped
        {
            ItemSet * tempSet = new ItemSet();
            tempSet->insert(nMovie);
            MovieMapByDirector.insert(std::make_pair(director, tempSet));
            ItemMapByTitle.insert(std::make_pair(title, nMovie));
        }
        else // Case 2 -  already mapped
        {
            ItemSet * tempSet;
            tempSet = MovieMapByDirector[director];
            tempSet->insert(nMovie);
            MovieMapByDirector.insert(std::make_pair(director, tempSet));
            ItemMapByTitle.insert(std::make_pair(title, nMovie));
        }

    return nMovie;
}

void Library::addCastMembers(const Item* const movie, const int nCastMembers, ...)
{
    // your code here

    va_list cast;
    char *castMember;

    va_start(cast, nCastMembers);

    for (int i = 0; i < nCastMembers; i++)
    {
	    castMember = va_arg(cast, char*);

	    // do something with each keyword
        
        Movie * nMovie;

        map<string, ItemPtr>::iterator it;
        it = ItemMapByTitle.find(movie->getTitle());
        nMovie = (Movie*)it->second.getPtr(); // Temp fix

        nMovie->setCast(castMember);

        if (MovieMapByCast.find(castMember) == MovieMapByCast.end()) // Case 1 - keyword not mapped
        {
            ItemSet * tempSet = new ItemSet();
            
            tempSet->insert(nMovie);
            MovieMapByCast.insert(std::make_pair(castMember, tempSet));
        }
        else // Case 2 - keyword already mapped
        {
            ItemSet * tempSet = MovieMapByCast[castMember];

            tempSet->insert(nMovie);

            MovieMapByCast.insert(std::make_pair(castMember, tempSet));
        }

    }

    va_end(cast);
}

const ItemSet* Library::moviesByDirector(const string& director) const
{
    // your code here
    const ItemSet * tempSet;

    if (MovieMapByDirector.find(director) != MovieMapByDirector.end())
    {
        tempSet = MovieMapByDirector.at(director);

        return tempSet;
    }
    else
    {
        return NULL;
    }
}

const ItemSet* Library::moviesByActor(const string& actor) const
{
    const ItemSet * tempSet;

    if (MovieMapByCast.find(actor) != MovieMapByCast.end())
    {
        tempSet = MovieMapByCast.at(actor);

        return tempSet;
    }
    else
    {
        return NULL;
    }
    return NULL;
}

const ItemSet* Library::movies() const
{
    // your code here

    return &MovieSet;
}

static void deleteMapContents(StringToItemSetMap& s2ism)
{
    // your code here

    // Iterate through map
    for(std::map<string, ItemSet*>::iterator it = s2ism.begin(); it != s2ism.end(); )
    {
        // Iterate through set
        ItemSet *tempSet = it->second;

        for(std::set<ItemPtr>::iterator it2 = tempSet->begin(); it2 != tempSet->end(); )
        {
          
            if (next((it2),1) != tempSet->end())        
                it2 = tempSet->erase(it2);
            else
                ++it2;
        }

//        tempSet->clear();
          delete tempSet; // Test, put in item constructor?

        if (next((it),1) != s2ism.end())        
            it = s2ism.erase(it);
        else
            ++it;
    }
     
//    s2ism.clear();

}

static void deleteItemSetContents(ItemSet& itemSet)
{
    // your code here

    for(std::set<ItemPtr>::iterator it = itemSet.begin(); it != itemSet.end(); )
    {
        delete it->getPtr();

        if (next((it),1) != itemSet.end())        
            it = itemSet.erase(it);
        else
            ++it;
    }

//    itemSet.clear();
}

File: CS261/mylibrary_java/library/Library.java
package library;

import java.io.PrintStream;
import java.util.Collection;
import java.util.*;

public class Library
{
    Map<String, TreeMap<String, Item>> albumsByBand = new TreeMap<>();
    Map<String, TreeMap<String, Item>> albumsByMember = new TreeMap<>();
    Map<String, Item> albumsByTitle = new TreeMap<>();
    Map<String, TreeMap<String, Item>> booksByAuthor = new TreeMap<>();
    Map<String, Item> booksByTitle = new TreeMap<>();
    Map<String, TreeMap<String, Item>> itemsByKeyword = new TreeMap<>();
    Map<String, TreeMap<String, Item>> moviesByActor = new TreeMap<>();
    Map<String, TreeMap<String, Item>> moviesByDirector = new TreeMap<>();
    Map<String, Item> moviesByTitle = new TreeMap<>();

	// general methods
	
	// returns all of the items which have the specified keyword
	public Collection<Item> itemsForKeyword(String keyword)
	{
        boolean debug = false;

        TreeMap<String, Item> tempMap = new TreeMap<>();
	
        if (debug == true)
        {
            System.out.println("itemsForKeyword ");
            System.out.println("keyword = " + keyword);
        }

        if (itemsByKeyword.containsKey(keyword))
        {
            if (debug == true)
                System.out.println(keyword + " found\n");

            tempMap = itemsByKeyword.get(keyword);
        }
        else 
        {
            if (debug == true)
                System.out.println(keyword + " not found\n");
        }

		return tempMap.values();
	}
	
	// print an item from this library to the output stream provided
	public void printItem(PrintStream out, Item item)
	{
        boolean debug = false;

        if (debug == true)
        {
            System.out.println("printItem ");
        }

        System.out.println(item.toString());
	}
	
	// book-related methods
	
	// adds a book to the library
	public Item addBook(String title, String author, int nPages, String... keywords)
	{
        boolean debug = false;

        if (debug == true)
        {
            System.out.println("addBook ");
            System.out.println("title = " + title);
            System.out.println("author = " + author);
            System.out.println("nPages = " + nPages);
            for(String n : keywords)
                System.out.println("keyword = " + n);
        }

        Item nBook = new Book(); 
        nBook.setTitle(title);
        nBook.setPerson(author);
        nBook.setNumberItems(nPages);
        nBook.setKeywords(keywords);

        for(String n : keywords)
        {
            if (!itemsByKeyword.containsKey(n))
            {
                TreeMap<String, Item> nMap = new TreeMap<>();
                nMap.put(title, nBook);

                itemsByKeyword.put(n, nMap);
            }
            else
            {
                TreeMap<String, Item> nMap = itemsByKeyword.get(n);
                nMap.put(title, nBook);

                itemsByKeyword.put(n, nMap);
            }
        }

        if (!booksByAuthor.containsKey(author))
        {
            TreeMap<String, Item> nMap = new TreeMap<>();
            nMap.put(title, nBook);

            booksByAuthor.put(author, nMap);
        }
        else
        {
            TreeMap<String, Item> nMap = booksByAuthor.get(author);
            nMap.put(title, nBook);

            booksByAuthor.put(author, nMap);
 
        }

        booksByTitle.put(title, nBook);
           
        if (debug == true)
        {
            System.out.println(nBook.toString());
        }

		return nBook;
	}
	
	// removes a book from the library
	public boolean removeBook(String title)
	{
        boolean debug = false;

        if (debug == true)
        {
            System.out.println("removeBook ");
            System.out.println("title = " + title + "\n");
        }

        if (booksByTitle.containsKey(title))
        {
            if (debug == true)
                System.out.println(title + " found");
            
            Item tempBook = booksByTitle.get(title);
            String[] keywords = tempBook.getKeywords();
            String author = tempBook.getPerson();

            if (debug == true)
            {
                System.out.println("title = " + tempBook.getTitle());
                for(String keyword : keywords )
                    System.out.println("keyword = " + keyword);
                System.out.println("author = " + author);
            }
           
            for(String keyword : keywords)
            {
                if(itemsByKeyword.containsKey(keyword))
                {
                    if (debug == true)
                        System.out.println(keyword + " found");

                    TreeMap<String, Item> tempTree = itemsByKeyword.get(keyword);

                    Item tempItem = tempTree.get(title);
                    tempTree.remove(title);
                }
                else
                {
                    if (debug == true)
                        System.out.println(keyword + " not found");
                }
            }

            if(booksByAuthor.containsKey(author))
            {
                if (debug == true)
                    System.out.println(author + " found");

                TreeMap<String, Item> tempTree = booksByAuthor.get(author);

                tempTree.clear();

            }
            else
            {
                if (debug == true)
                    System.out.println(author + " not found");
            }

            booksByTitle.remove(title);

            return true;
        }
        else
        {
            if (debug == true)
                System.out.println(title + " not found");

		    return false;
        }
	}
	
	// returns all of the books by the specified author
	public Collection<Item> booksByAuthor(String author)
	{
        TreeMap<String, Item> tempMap = new TreeMap<>();

        boolean debug = false;

        if (debug == true)
        {
            System.out.println("booksByAuthor ");
            System.out.println("author = " + author);
        }

        if (booksByAuthor.containsKey(author))
        {
            if (debug == true)
                System.out.println(author + " found\n");

            tempMap = booksByAuthor.get(author);
        }
        else 
        {
            if (debug == true)
                System.out.println(author + " not found\n");
        }

		return tempMap.values();
	}
	
	// returns all of the books in the library
	public Collection<Item> books()
	{
        boolean debug = false;

        if (debug == true)
        {
            System.out.println("books ");
        }

		return booksByTitle.values();
	}
	
	// music-related methods
	
	// adds a music album to the library
	public Item addMusicAlbum(String title, String band, int nSongs, String... keywords)
	{
        boolean debug = false;

        if (debug == true)
        {
            System.out.println("addMusicAlbum ");
            System.out.println("title = " + title);
            System.out.println("band = " + band);
            System.out.println("nSongs = " + nSongs);
            for(String n : keywords)
                System.out.println("keyword = " + n);
        }

        Item nAlbum = new MusicAlbum(); 
        nAlbum.setTitle(title);
        nAlbum.setPerson(band);
        nAlbum.setNumberItems(nSongs);
        nAlbum.setKeywords(keywords);

        for(String n : keywords)
        {
            if (!itemsByKeyword.containsKey(n))
            {
                TreeMap<String, Item> nMap = new TreeMap<>();
                nMap.put(title, nAlbum);

                itemsByKeyword.put(n, nMap);
            }
            else
            {
                TreeMap<String, Item> nMap = itemsByKeyword.get(n);
                nMap.put(title, nAlbum);

                itemsByKeyword.put(n, nMap);
            }
        }

        if (!albumsByBand.containsKey(band))
        {
            TreeMap<String, Item> nMap = new TreeMap<>();
            nMap.put(title, nAlbum);

            albumsByBand.put(band, nMap);
        }
        else
        {
            TreeMap<String, Item> nMap = albumsByBand.get(band);
            nMap.put(title, nAlbum);

            albumsByBand.put(band, nMap);
        }

        albumsByTitle.put(title, nAlbum);

        if (debug == true)
        {
            System.out.println(nAlbum.toString());
        }

		return nAlbum;
	}

	// adds the specified band members to a music album
	public void addBandMembers(Item album, String... members)
	{
        boolean debug = false;

        if (debug == true)
        {
            System.out.println("addBandMembers ");
            System.out.println("album = " + album.getTitle());
            for(String n : members)
                System.out.println("member = " + n);
            System.out.println("");
        }

        MusicAlbum nAlbum = (MusicAlbum) album;
        nAlbum.setMembers(members);
            
        for (String member : members) 
        {
            if (!albumsByMember.containsKey(member))
            {
                TreeMap<String, Item> nMap = new TreeMap<>();
                nMap.put(nAlbum.getTitle(), nAlbum);

                albumsByMember.put(member, nMap);
            }
            else
            {
                TreeMap<String, Item> nMap = albumsByMember.get(member);
                nMap.put(nAlbum.getTitle(), nAlbum);

                albumsByMember.put(member, nMap);
            }
        }
	}
	
	// removes a music album from the library
	public boolean removeMusicAlbum(String title)
	{
        boolean debug = false;

        if (debug == true)
        {
            System.out.println("removeMusicAlbum ");
            System.out.println("title = " + title);
        }

        if (albumsByTitle.containsKey(title))
        {
            if (debug == true)
                System.out.println(title + " found");
            
            MusicAlbum tempAlbum = (MusicAlbum) albumsByTitle.get(title);
            String band = tempAlbum.getPerson();
            String[] keywords = tempAlbum.getKeywords();
            String[] members = tempAlbum.getMembers();

            if (debug == true)
            {
                System.out.println("title = " + tempAlbum.getTitle());
                for(String keyword : keywords)
                    System.out.println("keyword = " + keyword);
                System.out.println("band = " + band);
                
            }
           
            for(String keyword : keywords)
            {
                if(itemsByKeyword.containsKey(keyword))
                {
                    if (debug == true)
                        System.out.println(keyword + " found");

                    TreeMap<String, Item> tempTree = itemsByKeyword.get(keyword);

                    Item tempItem = tempTree.get(title);
                    tempTree.remove(title);
                }
                else
                {
                    if (debug == true)
                        System.out.println(keyword + " not found");
                }
            }

            if(albumsByBand.containsKey(band))
            {
                if (debug == true)
                    System.out.println(band + " found");

                TreeMap<String, Item> tempTree = albumsByBand.get(band);

                Item tempItem = tempTree.get(title);
                tempTree.remove(title);

            }
            else
            {
                if (debug == true)
                    System.out.println(band + " not found");
            }

            for(String member : members)
            {
                if(albumsByMember.containsKey(member))
                {
                    if (debug == true)
                        System.out.println(member + " found");

                    TreeMap<String, Item> tempTree = albumsByMember.get(member);

                    Item tempItem = tempTree.get(title);
                    tempTree.remove(title);
                }
                else
                {
                    if (debug == true)
                        System.out.println(member + " not found");
                }
            }

            albumsByTitle.remove(title);

            return true;
        }
        else
        {
            if (debug == true)
                System.out.println(title + " not found");

		    return false;
        }
	}

	// returns all of the music albums by the specified band
	public Collection<Item> musicByBand(String band)
	{
        TreeMap<String, Item> tempMap = new TreeMap<>();

        boolean debug = false;

        if (debug == true)
        {
            System.out.println("musicByBand ");
            System.out.println("band = " + band);
        }

        if (albumsByBand.containsKey(band))
        {
            if (debug == true)
                System.out.println(band + " found\n");

            tempMap = albumsByBand.get(band);
        }
        else 
        {
            if (debug == true)
                System.out.println(band + " not found\n");
        }

		return tempMap.values();
	}
	
	// returns all of the music albums by the specified musician
	public Collection<Item> musicByMusician(String musician)
	{
        TreeMap<String, Item> tempMap = new TreeMap<>();

        boolean debug = false;

        if (debug == true)
        {
            System.out.println("musicByMusician ");
            System.out.println("musician = " + musician);
        }

        if (albumsByMember.containsKey(musician))
        {
            if (debug == true)
                System.out.println(musician + " found\n");

            tempMap = albumsByMember.get(musician);
        }
        else 
        {
            if (debug == true)
                System.out.println(musician + " not found\n");
        }

		return tempMap.values();

	}
	
	// returns all of the music albums in the library
	public Collection<Item> musicAlbums()
	{
        boolean debug = false;

        if (debug == true)
        {
            System.out.println("musicAlbums ");
        }

		return albumsByTitle.values();
	}
	
	// movie-related methods
	
	// adds a movie to the library
	public Item addMovie(String title, String director, int nScenes, String... keywords)
	{
        boolean debug = false;

        if (debug == true)
        {
            System.out.println("addMovie ");
            System.out.println("title = " + title);
            System.out.println("director = " + director);
            System.out.println("nScenes = " + nScenes);
            for(String n : keywords)
                System.out.println("keyword = " + n);
        }

        Item nMovie = new Movie(); 
        nMovie.setTitle(title);
        nMovie.setPerson(director);
        nMovie.setNumberItems(nScenes);
        nMovie.setKeywords(keywords);

        for(String n : keywords)
        {
            if (!itemsByKeyword.containsKey(n))
            {
                TreeMap<String, Item> nMap = new TreeMap<>();
                nMap.put(title, nMovie);

                itemsByKeyword.put(n, nMap);
            }
            else
            {
                TreeMap<String, Item> nMap = itemsByKeyword.get(n);
                nMap.put(title, nMovie);

                itemsByKeyword.put(n, nMap);
            }
        }

        if (!moviesByDirector.containsKey(director))
        {
            TreeMap<String, Item> nMap = new TreeMap<>();
            nMap.put(title, nMovie);

            moviesByDirector.put(director, nMap);
        }
        else
        {
            TreeMap<String, Item> nMap = moviesByDirector.get(director);
            nMap.put(title, nMovie);

            moviesByDirector.put(director, nMap);
        }
        moviesByTitle.put(title, nMovie);

        if (debug == true)
        {
            System.out.println(nMovie.toString());
        }

		return nMovie;
	}

	// adds the specified actors to a movie
	public void addCast(Item movie, String... members)
	{
        boolean debug = false;

        if (debug == true)
        {
            System.out.println("addCast ");
            System.out.println("movie = " + movie.getTitle());
            for(String n : members)
                System.out.println("member = " + n);
            System.out.println("");
        }

        Movie nMovie = (Movie) movie;
        nMovie.setCast(members);

        for (String member : members) 
        {
            if (!moviesByActor.containsKey(member))
            {
                TreeMap<String, Item> nMap = new TreeMap<>();
                nMap.put(nMovie.getTitle(), nMovie);

                moviesByActor.put(member, nMap);
            }
            else
            {
                TreeMap<String, Item> nMap = moviesByActor.get(member);
                nMap.put(nMovie.getTitle(), nMovie);

                moviesByActor.put(member, nMap);
            }
        }
	}

	// removes a movie from the library
	public boolean removeMovie(String title)
	{
        boolean debug = false;

        if (debug == true)
        {
            System.out.println("removeMovie ");
            System.out.println("title = " + title);
        }

        if (moviesByTitle.containsKey(title))
        {
            if (debug == true)
                System.out.println(title + " found");
            
            Movie tempMovie = (Movie) moviesByTitle.get(title);
            String[] cast = tempMovie.getCast();
            String director  = tempMovie.getPerson();
            String[] keywords = tempMovie.getKeywords();

            if (debug == true)
            {
                System.out.println("title = " + tempMovie.getTitle());
                for(String keyword : keywords)
                    System.out.println("keyword = " + keyword);
                System.out.println("director = " + director);
                
            }
           
            for(String keyword : keywords)
            {
                if(itemsByKeyword.containsKey(keyword))
                {
                    if (debug == true)
                        System.out.println(keyword + " found");

                    TreeMap<String, Item> tempTree = itemsByKeyword.get(keyword);

                    Item tempItem = tempTree.get(title);
                    tempTree.remove(title);
                }
                else
                {
                    if (debug == true)
                        System.out.println(keyword + " not found");
                }
            }

            if(moviesByDirector.containsKey(director))
            {
                if (debug == true)
                    System.out.println(director + " found");

                TreeMap<String, Item> tempTree = moviesByDirector.get(director);

                Item tempItem = tempTree.get(title);
                tempTree.remove(title);

            }
            else
            {
                if (debug == true)
                    System.out.println(director + " not found");
            }

            for(String actor : cast)
            {
                if(moviesByActor.containsKey(actor))
                {
                    if (debug == true)
                        System.out.println(actor + " found");

                    TreeMap<String, Item> tempTree = moviesByActor.get(actor);

                    Item tempItem = tempTree.get(title);
                    tempTree.remove(title);
                }
                else
                {
                    if (debug == true)
                        System.out.println(actor + " not found");
                }
            }

            moviesByTitle.remove(title);

            return true;
        }
        else
        {
            if (debug == true)
                System.out.println(title + " not found");

		    return false;
        }
	}
	
	// returns all of the movies by the specified director
	public Collection<Item> moviesByDirector(String director)
	{
        TreeMap<String, Item> tempMap = new TreeMap<>();

        boolean debug = false;

        if (debug == true)
        {
            System.out.println("moviesByDirector ");
            System.out.println("director = " + director);
        }

        if (moviesByDirector.containsKey(director))
        {
            if (debug == true)
                System.out.println(director + " found\n");

            tempMap = moviesByDirector.get(director);
        }
        else 
        {
            if (debug == true)
                System.out.println(director + " not found\n");
        }

		return tempMap.values();
	}
	
	// returns all of the movies by the specified actor
	public Collection<Item> moviesByActor(String actor)
	{
        TreeMap<String, Item> tempMap = new TreeMap<>();

        boolean debug = false;

        if (debug == true)
        {
            System.out.println("moviesByActor ");
            System.out.println("actor = " + actor);
        }

        if (moviesByActor.containsKey(actor))
        {
            if (debug == true)
                System.out.println(actor + " found\n");

            tempMap = moviesByActor.get(actor);
        }
        else 
        {
            if (debug == true)
                System.out.println(actor + " not found\n");
        }

		return tempMap.values();
	}
	
	// returns all of the movies in the library
	public Collection<Item> movies()
	{
        boolean debug = false;

        if (debug == true)
        {
            System.out.println("movies ");
        }

		return moviesByTitle.values();
	}	
}
